### PR TITLE
HPCC-34007 null could be unintentionally set in a JSON string

### PR DESCRIPTION
**Pull request by streeterd**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — APPROVED** [review]
This is a useful macro that appears at the head of many of our test files:

`#define CPPUNIT_ASSERT_EQUAL_STR(x, y) CPPUNIT_ASSERT_EQUAL(std::string(x ? x : ""),std::string(y ? y : ""))
`

**Comment by streeterd** [comment]
> This is a useful macro that appears at the head of many of our test files:
> 
> `#define CPPUNIT_ASSERT_EQUAL_STR(x, y) CPPUNIT_ASSERT_EQUAL(std::string(x ? x : ""),std::string(y ? y : "")) `

Fixed in 57faf656e59f501f35a0bf842c94cc028727fe90

**Comment by streeterd** [comment]
@ghalliday  @jakesmith  @AttilaVamos  please review.

File: testing/unittests/commonJsonWriterTests.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+    void testArrays()
+    {
+        CommonJsonWriter writer(0);
+
+        // Test array output
+        writer.outputBeginArray("numbers");
+
+        // Output some array items
+        writer.outputInt(1, sizeof(int), NULL);
```
Comment (jakesmith) [inline testing/unittests/commonJsonWriterTests.cpp:123]: trivial: nullptr vs NULL prefered

**Review by jakesmith — COMMENTED** [review]
@streeterd - a couple of minor comments.

It would be good if these unit tests, also tested the original problem (HPCC-33963), i.e. outputting an empty unicode string.

**Review by ghalliday — APPROVED** [review]

File: testing/unittests/commonJsonWriterTests.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    void setUp() {}
+    void tearDown() {}
+
+    void testNullValues()
+    {
+        // Test with default flags (no indentation)
+        CommonJsonWriter writer(XWFnoindent);
+
+        // Test null string output
+        char *pChar = nullptr;
```
[the remark this answers is not in the document]
Comment (streeterd) [reply testing/unittests/commonJsonWriterTests.cpp]: Fixed in 95c86aabf047acfc148af445021083072490ee81

File: testing/unittests/commonJsonWriterTests.cpp (hunk 1 above)
Comment (streeterd) [reply testing/unittests/commonJsonWriterTests.cpp]: Fixed in 95c86aabf047acfc148af445021083072490ee81

**Comment by streeterd** [comment]
> @streeterd - a couple of minor comments.
> 
> It would be good if these unit tests, also tested the original problem (HPCC-33963), i.e. outputting an empty unicode string.

@jakesmith this unit test is designed to test the class CommonJsonWriter. The HPCC-33963 issue was in rtl/eclrtl/rtlxml.cpp. I will add a unit test for the rtlxml functions in a separate unit test.

**Comment by jakesmith** [comment]
> @streeterd to clarify for the future: if I approve a PR but also add a comment, there is no need to make the change for that PR (and quite possibly for other PRs). For something like this I am happy with the change, but would probably prefer the change in a later PR so this change is not held up. @jakesmith do you agree? This also has another comment from Jake, so it may have needed changes anyway.

yes, agree. It is better to create a fix PR and separate PR(s) for unit tests that are missing.
If introducing a new feature, it is better to introduce the unit tests at the same time as the new code..

**Comment by jakesmith** [comment]
> > @streeterd - a couple of minor comments.
> > It would be good if these unit tests, also tested the original problem (HPCC-33963), i.e. outputting an empty unicode string.
> 
> @jakesmith this unit test is designed to test the class CommonJsonWriter. The HPCC-33963 issue was in rtl/eclrtl/rtlxml.cpp. I will add a unit test for the rtlxml functions in a separate unit test.

It would still be good if these tests test a blank string too not just null. It won't test the HPCC-33963 fix, but theoretically it's an edge case with the json writer too..

**Review by AttilaVamos — APPROVED** [review]
Seems good to go.

**Comment by streeterd** [comment]
> > > @streeterd - a couple of minor comments.
> > > It would be good if these unit tests, also tested the original problem (HPCC-33963), i.e. outputting an empty unicode string.
> > 
> > 
> > @jakesmith this unit test is designed to test the class CommonJsonWriter. The HPCC-33963 issue was in rtl/eclrtl/rtlxml.cpp. I will add a unit test for the rtlxml functions in a separate unit test.
> 
> It would still be good if these tests test a blank string too not just null. It won't test the HPCC-33963 fix, but theoretically it's an edge case with the json writer too..

Added in 857818d62d445efad0f497949241dfc7b31bfa87

**Review by AttilaVamos — COMMENTED** [review]
One question.

File: testing/unittests/rtlxmltests.cpp [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+        CPPUNIT_ASSERT_EQUAL_STR("hello", out.str());
+        out.clear();
+
+        // Test with special characters
+        outputXmlString(15, "hello & <world>", "greeting", out);
+        CPPUNIT_ASSERT_EQUAL_STR("<greeting>hello &amp; &lt;world&gt;</greeting>", out.str());
+    }
+
+    void testXmlBool()
+    {
```
Comment (AttilaVamos) [inline testing/unittests/rtlxmltests.cpp:67]: What about to testing empty string and nullptr (like in commonJsonWriterTests)?

File: testing/unittests/rtlxmltests.cpp [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    CPPUNIT_TEST(testJsonDecimal);
+    CPPUNIT_TEST(testComplexStructure);
+    CPPUNIT_TEST_SUITE_END();
+
+public:
+    void setUp() {}
+    void tearDown() {}
+
+    void testXmlString()
+    {
```
Comment (AttilaVamos) [inline testing/unittests/rtlxmltests.cpp:48]: What about to testing empty string and nullptr (like in commonJsonWriterTests)?

Comment (streeterd) [reply testing/unittests/rtlxmltests.cpp:48]: Added in e8241faecd88e6b38974feb8519fd61f19eb9059

**Review by ghalliday — APPROVED** [review]
Looks good.  Please squash.

**Comment by streeterd** [comment]
@ghalliday squashed.

File: testing/unittests/rtlxmltests.cpp [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+
+        // Test with null fieldname
+        outputXmlString(5, "hello", nullptr, out);
+        CPPUNIT_ASSERT_EQUAL_STR("hello", out.str());
+        out.clear();
+
+        // Test with empty string fieldname
+        outputXmlString(5, "hello", "", out);
+        CPPUNIT_ASSERT_EQUAL_STR("hello", out.str());
+        out.clear();
```
Comment (AttilaVamos) [inline testing/unittests/rtlxmltests.cpp:64]: Is there a sense to add same empty string and nullptr test for 'field' parameter? (And both 'field' and 'fieldname')?

**Review by AttilaVamos — COMMENTED** [review]
One more question.

**Review by jakesmith — APPROVED** [review]
@streeterd - looks good.
